source/beamline/CLS/CLSShutters.cpp

```diff
@@ -14,7 +19,88 @@ CLSShutters::CLSShutters(const QString &name, QObject *parent) :
 
 CLSShutters::~CLSShutters()
 {
+	shuttersBeamOnOrderMap_.clear();
+}
+
+AMListAction3* CLSShutters::createBeamOnActionList()
+{
+	// create the action list to move the Shutters (sequentially) and wait for the move done
+	AMListAction3 *openShuttersActionList = new AMListAction3(new AMListActionInfo3("Open shutters action list", "Open Shutters"), AMListAction3::Sequential);
+	AMListAction3 *waitShuttersOpenActionList = new AMListAction3(new AMListActionInfo3("Wait shutters open action list", "Wait Shutters Open"), AMListAction3::Parallel);
+
+	// this is to make sure all the controls are checked
+	int currentBeamOnOrder = 1;
+	int checkedControlCount = 0;
+	while (checkedControlCount < shuttersBeamOnOrderMap_.count()) {
+		AMControl *shutterControl = shuttersBeamOnOrderMap_.value(currentBeamOnOrder);
+		if (shutterControl) {
+			if (isChildState2(shutterControl)) { // the Shutter is closed
+				double valveOpenValue = controlState1ValueMap_.value(shutterControl);
+				AMAction3 *openShutterAction = AMActionSupport::buildControlMoveAction(shutterControl, valveOpenValue);
+				openShuttersActionList->addSubAction(openShutterAction);
+
+				AMAction3 *waitShutterOpenAction = AMActionSupport::buildControlWaitAction(shutterControl, valveOpenValue);
+				waitShuttersOpenActionList->addSubAction(waitShutterOpenAction);
+			}
+
+			checkedControlCount ++;
+		}
+
+		currentBeamOnOrder++;
+	}
+
+	// add the open/wait action lists to the beam on action list
+	AMListAction3 *openShuttersActionsList = 0;
+	if (openShuttersActionList->subActionCount() > 0) {
+		openShuttersActionsList = new AMListAction3(new AMListActionInfo3("Beam On - Shutter", "Beam On: Open shutters"), AMListAction3::Parallel);
+		openShuttersActionsList->addSubAction(openShuttersActionList);
+		openShuttersActionsList->addSubAction(waitShuttersOpenActionList);
+	} else {
+		openShuttersActionList->deleteLater();
+		waitShuttersOpenActionList->deleteLater();
+	}
+
+	return openShuttersActionsList;
+}
+
+AMListAction3* CLSShutters::createBeamOffActionList()
+{
+	// create the action list to move the Shutters (sequentially) and wait for the move done
+	AMListAction3 *closeShuttersActionList = new AMListAction3(new AMListActionInfo3("Close shutters action list", "Close Shutters"), AMListAction3::Sequential);
+	AMListAction3 *waitShuttersCloseActionList = new AMListAction3(new AMListActionInfo3("Wait shutters closed action list", "Wait Shutters Closed"), AMListAction3::Parallel);
+
+	// this is to make sure all the controls are checked
+	int currentBeamOffOrder = 1;
+	int checkedControlCount = 0;
+	while (checkedControlCount < shuttersBeamOnOrderMap_.count()) {
```

Comment:
it is still in order of map keys. The next line of codes gets the shutter in the order. This `while` loop will make sure we go through all the ordered shutters